components/add.js

```diff
@@ -0,0 +1,331 @@
+import { useRouter } from "next/router";
+import { empty } from "empty-schema";
+import { PROFILE_COLLECTIONS } from "../lib/profiles";
+
+/**
+ *
+ * Given the schema ID, fetch the corresponding collection
+ * name. Example: schema id /profiles/human_donor.json maps
+ * to "human-donors"
+ */
+export const collectionPath = (schema) => {
+  return PROFILE_COLLECTIONS[schema["$id"]];
+};
+
+/**
+ * Looks for an action in the item with a name in the list of given actions
+ * and returns the profile path of the first one found, or null if none are found.
+ * @param {*} item snovault object
+ * @param {*} actions list of actions to find from which to grab
+ * the profile path
+ * @returns the profile path corresponding to an action
+ */
+const actionProfile = (item, actions) => {
+  if ("actions" in item) {
+    const act = item.actions.find((act) => actions.includes(act.name));
+    if (act != null) {
+      return act.profile;
+    } else {
+      return null;
+    }
+  } else {
+    return null;
+  }
+};
+
+/**
+ * This takes a schema object and adds to the required properties
+ * in order for the blank template library to autogenerate stubs
+ * for as many fields as makes sense
+ *
+ * We eliminate properties that have permission requirements, not
+ * submittable, commented with "Do not submit", or arrays/objects.
+ */
+const convertOptionalIntoRequiredSchema = (schema) => {
+  const topProperties = Object.entries(schema.properties)
+    .filter(([, p]) => {
+      return p.comment ? !p.comment.includes("Do not submit") : true;
+    })
+    .filter(([, p]) => {
+      return !p.permission;
+    })
+    .filter(([, p]) => {
+      return p.notSubmittable ? !p.notSubmittable : true;
+    })
+    .filter(([, p]) => {
+      return !["array", "object"].includes(p.type);
+    })
+    .map(([k]) => {
+      return k;
+    });
+  const newschema = schema;
+  newschema.required = [...topProperties, ...schema.required];
+  return newschema;
+};
+
+export const useEditor = (collection, viewComponent, action = "edit") => {
+  /**
+   * Represents whether the Editor component can be actively typed in or saved.
+   * This is determined by the logged in status of the user and if the user has
+   * edit permissions on that object being edited.
+   */
+  const [edit, setEditing] = useState(false);
+
+  const router = useRouter();
+
+  useEffect(() => {
+    const isEdit = document.URL.endsWith(`#!${action}`);
+    // If the URL says edit but we aren't editing yet, set the state
+    if (isEdit && !edit) {
+      setEditing(true);
+    }
+
+    // If the URL has us not editing but we just were, set to false, and update props
+    if (!isEdit && edit) {
+      setEditing(false);
+      router.replace(router.asPath);
+    }
+  }, [edit, router, action]);
+
+  const addpage = <AddInstancePage collection={collection} />;
+
+  const componentToShow = edit ? addpage : <>{viewComponent}</>;
+  return componentToShow;
```

Comment:
oh okay, should I just put this directly in <AddInstancePage> and have a children prop essentially? It would show if not `edit`, otherwise show the editor?

Reply:
Since you share this between add and edit, it seems fine to use this as a custom hook. But I think I would have it return `editing` and `setEditing` and have add and edit decide what JSX to render in response, whether the component unmodified, or the add/edit page. You’ll have more duplication, but very little. And if you did have a lot of duplication as a result of this, I would refactor the duplicate JSX into another component that wasn’t a custom hook.